simplesat/transaction.py

```diff
@@ -113,23 +121,40 @@ def _safe_operations(self, pool, decisions, installed_package_ids):
                       package_id not in installed_package_ids):
                     installs.append(package_id)
 
-        # Removals should happen top down
-        for package_id in reversed(removals):
-            package = pool.id_to_package(package_id)
-            operations.append(RemoveOperation(package))
+        install_operations = []
+        remove_operations = []
 
         # Installations should happen bottom up
         for package_id in installs:
-            package = pool.id_to_package(package_id)
-            operations.append(InstallOperation(package))
+            # The solver may suggest removing a currently installed package and
+            # installing another version of that same package from a different
```

Comment:
At the moment, we don't exactly "handle" this. We do treat them as equivalent: the policy and solver don't care about the repository info, just the package name and version. In a discussion on [Slack](https://enthought.slack.com/archives/edm/p1464364546004725) @cournape noted, "Even if we implement repository priority at a latter time, we will still always prefer installed over non installed for a same `(name, version)` pair."

https://github.com/enthought/sat-solver/pull/220/commits/b8aedbb9376103b39886da7a04fc89f79b76f7ba describes the reason for this change. It's not the only way to do this but it was easier than overhauling more of the policy to escape the infinite loop.